Make C++ validation tests run on all platforms

var-const opened the pull request:
Also:
* updated the GoogleTest version (to be able to use the `Throws` matcher);
* enabled exceptions in Firestore integration tests (otherwise validation tests would be ignored);
* add some missing validation (see the linked bug for remaining gaps);
* replace all usage of `FIREBASE_ASSERT` macros with `SIMPLE_HARD_ASSERT` (which throws an exception on Unity / crash in C++).

See b/136012313 for more context.

dconeybe commented on firestore/integration_test_internal/src/validation_test.cc:
```diff
@@ -19,27 +19,276 @@
+             "any other methods on a Firestore instance.";
+#endif
+
+    case ErrorCase::kSettingsDisableSsl:
```
The two values for `kSettingsDisableSsl` are identical. Can you get rid of the `#ifdef` for this one?

var-const replied:
Thanks for spotting this! I think I originally thought that the C++ validation message would spell it `ssl_enabled`, but it turned out to be not true.

dconeybe commented on firestore/integration_test_internal/src/validation_test.cc:
```diff
@@ -120,49 +364,24 @@ class ValidationTest : public FirestoreIntegrationTest {
+    EXPECT_ERROR_EITHER(
+        document.Update(MapFieldValue{{path, FieldValue::Integer(1)}}), reason,
+        // TODO(b/171990785): Unify Android and C++ validation error messages.
+        // When validated by Android Java code
```
This comment `When validated by Android Java code` doesn't make sense to me. Is it a copy-paste error? If not, could you re-write it as a full English sentence?

var-const replied:
Done. I think the original idea was something like "The message below is produced when validation is performed by the Android SDK", hopefully it reads a little better now.

dconeybe commented on firestore/integration_test_internal/src/validation_test.cc:
```diff
@@ -350,8 +515,12 @@ TEST_F(ValidationTest, WritesMayContainIndirectlyNestedLists) {
-// TODO(zxu): There is no way to create Firestore with different project id yet.
-TEST_F(ValidationTest, WritesMustNotContainReferencesToADifferentDatabase) {}
+// TODO(b/136012313): batches and transactions don't currently throw.
+// TEST_F(ValidationTest, WritesMustNotContainReferencesToADifferentDatabase) {
```
Consider uncommenting this test, but skipping it via the `GTEST_SKIP()` macro. This will at least highlight in the test result count that a test has been skipped. Leaving it commented here makes it less visible.

There is precedence for this too, like all of the macros in https://github.com/firebase/firebase-cpp-sdk/blob/8ad99ec49ff5d01e909d5cedc5366cf5325460ec/testing/test_framework/src/firebase_test_framework.h.

var-const replied:
Thanks, great suggestion.

dconeybe commented on firestore/src/main/transaction_main.cc:
```diff
@@ -156,5 +163,14 @@ void TransactionInternal::MarkPermanentlyFailed() {
 }
 
+void TransactionInternal::ValidateReference(const DocumentReference& document) {
+  auto* internal_doc = GetInternal(&document);
```
Isn't it possible that the "internal" DocumentReference is nullptr? If so, you should probably check if it's null before using it.

var-const replied:
Good catch. The existing code would crash in that case. I'll add a validation test to cover this tomorrow.

dconeybe replied:
Just waiting for this to be addressed. Then I'll approve.

var-const replied:
Added the test (also addressed the same problem with `WriteBatch`).

dconeybe reviewed (changes requested):
I think you should add a change log entry, since there are externally-visible behavior changes being made by this PR.

var-const commented:
Some changes since the last review:
* added validation and tests for the case where an invalid `DocumentReference` is passed to a `Transaction` or a `WriteBatch` (previously it would crash);
* made `WritesMustNotContainReferencesToADifferentDatabase` work by adding a new test helper function, `TestFirestoreWithProjectId` (this necessitated a slight refactoring in `FirestoreIntegrationTest`);
* updated the GoogleTest version to be able to use the `Throws` matcher;
* enabled exceptions in Firestore integration tests;
* added a `README` entry.

dconeybe commented on firestore/src/common/document_reference.cc:
```diff
@@ -113,12 +114,23 @@ CollectionReference DocumentReference::Parent() const {
+  if (!collection_path) {
+    SimpleThrowInvalidArgument("Collection path cannot be null.");
+  }
+  if (collection_path[0] == '\0') {
```
Is there a reason to not use `strlen(collection_path) == 0` here, instead of `collection_path[0] == '\0'`?

var-const replied:
I don't have a strong preference on this -- it just seemed a little simpler to do.

dconeybe replied:
Personally, I'd prefer `strlen` because IMO it is more readable; that is, it directly conveys what you are checking. Checking that the first element == 0 is indirectly checking the string's length. This won't block my approval though.

dconeybe commented on firestore/integration_test_internal/CMakeLists.txt:
```diff
@@ -378,5 +378,9 @@ add_subdirectory(${FIREBASE_CPP_SDK_DIR} bin/ EXCLUDE_FROM_ALL)
+# Validation tests need exceptions to be enabled in order to run. The only
+# library that explicitly disables exceptions is App (but it gets propagated
+# everywhere), so overriding that effectively reenables exceptions.
+target_compile_options(firebase_app PUBLIC -fexceptions)
```
FYI This change probably won't be required once https://github.com/firebase/firebase-cpp-sdk/pull/561 is merged. Can you see if you can undo this change if you merge after it?

var-const replied:
Thanks for the pointer! I'll give it a try once that PR is merged.

jonsimantov commented:
Has the latest _main_ been merged into this PR, to ensure that the GoogleTest update doesn't break the flaky-test-workaround code in test_framework?

var-const commented on firestore/integration_test_internal/src/validation_test.cc:
```diff
@@ -172,44 +381,30 @@ TEST_F(ValidationTest, ChangingSettingsAfterUseFails) {
+}
 
 TEST_F(ValidationTest,
        FirestoreGetInstanceWithNonNullAppReturnsNonNullInstance) {
```
Added a null pointer check as well.

var-const commented on firestore/integration_test_internal/src/validation_test.cc:
```diff
@@ -381,40 +545,107 @@ TEST_F(ValidationTest, WritesMustNotContainReservedFieldNames) {
+  MapFieldValue data{{"foo", FieldValue::Integer(1)}};
+  DocumentReference bad_ref = db2->Document("foo/bar");
+
+#if defined(__ANDROID__)
```
Added a TODO.

var-const commented:
@jonsimantov My branch had some of those, but not all of them (e.g. I had some `installations` tests with flaky sections, but not `storage` tests). Just to be sure, I merged the latest `main` and will run all tests again.

jonsimantov reviewed (approved):
LGTM for the googletest changes. Deferring to Denver's review for the Firestore-specific test changes.
